### DISCUSSION TITLE
Derive on partial state

### DISCUSSION THREAD

**Discussion by ndrean** [pmndrs/valtio#502] [Q&A] [answered]
Is there a way to derive from a state but remove some dependencies, some partial state?
For example, below, the state depends on "id" and "user". If I make a derivation whose result depends only upon "id", is it possible to remove the dependency on "user"? Because any change to the "user" prop will make the derivation rerender, which I don't want.

My state is
```js
const store = ({
  id: 1,
  user: "me@mail.com",
  comments: []
})
```

I populate the "comments" on the fly via an async fetch, say:

```js
const fetchComments = (id) => fetch(`url/${id}`).then(res => res.json())
```

To make the comments field  reactive to the "id" field change, I believe I need to use a derivation:

```js
const derComments = derive({
  live: async get => {
    const list = await fetchComments(get(store).id)
   return list?.map( c => c.email)
```

My component is:
```js
const Component = ({store, derStore}) => {
    const { id } = useSnapshot(store)
   const { live } = useSnapshot(derStore)

   return(
    <>
      <button onClick={() => store.id += 1)>Increment</button>
     <p>{live}</p>
  </>
)
```

So that I need to use "suspense":

```js
<React.Suspense fallback={"Loading..."}>
   <Component store={store} derStore={derComments}/>
</React.Suspense>
```

This "works", the component updates (without a `useEffect`). 

However, the field "user" belongs to the store, and when I change it, then Component rerenders.

I could remove "user" from this state, split it and it's done but then I need to refactor the whole code.

**Comment by dai-shi** [comment, marked as answer]
> Is there a way to derive from a state but remove some dependencies, some partial state?

No, it's how `derive` is designed.

If you need fine-grained re-computation, you can try `proxyWithComputed` and `proxy-memoize`.

> I could remove "user" from this state, split it and it's done but then I need to refactor the whole code.

I would generally recommend it, more than `proxyWithComputed`.

```js
const store = ({
  id: { value: 1 },
  user: "me@mail.com",
  comments: []
})
```

**Reply by ndrean** [reply]
Ok, 
so we should atomize the state to remove interactions and thus  unnecessary  rendering. 

 `derive` is underestandable but `proxyWithComputed` seems complicated.

Now I am not sure why you changed `id: 1` to  `id: { value: 1 }`

**Reply by dai-shi** [reply]
If you change it to `id: { value: 1 }`, derive can be more atomic.

```js
const derComments = derive({
  live: async get => {
    const list = await fetchComments(get(store.id).value)
    return list?.map( c => c.email)
```

**Reply by ndrean** [reply]
`get` now can fine-grain indeed. Great saving

**Reply by ndrean** [reply]
The answer was [here](https://github.com/pmndrs/valtio/wiki/derive-util-usage-patterns), didn't found it! Sorry
